File: javascript/space-age/space_age.js
var SECONDS_IN_AN_EARTH_DAY = 24 * 60 * 60,
  EARTH_DAYS_IN_A_YEAR = {
    Mercury: 87.97,
    Earth: 365.25,
    Venus: 224.7,
    Mars: 687,
    Jupiter: 4331.86,
    Saturn: 10760.26,
    Uranus: 30706.57,
    Neptune: 60193.2
  };

function ageOn(planet, seconds) {
  var age = seconds / (EARTH_DAYS_IN_A_YEAR[planet] * SECONDS_IN_AN_EARTH_DAY);
  return +age.toFixed(2);
}

function SpaceAge(seconds) {
  this.seconds = seconds;
}

Object.keys(EARTH_DAYS_IN_A_YEAR).forEach(function (planetName) {
  SpaceAge.prototype['on' + planetName] = function () {
    return ageOn(planetName, this.seconds);
  };
});

module.exports = SpaceAge;

File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :neptune | :uranus

  @seconds_in_a_day 60 * 60 * 24

  @days_in_a_year [
    mercury: 87.9691,
    venus: 224.701,
    earth: 365.25,
    mars: 686.971,
    jupiter: 4_332.59,
    saturn: 10_759.22,
    neptune: 60_182,
    uranus: 30_688.5
  ]

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds),
  do: seconds / @seconds_in_a_day / @days_in_a_year[planet]
end
